agent/lib_monolog.c

```diff
@@ -0,0 +1,387 @@
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s: Logger has NULL class entry",
+                     __func__);
+    return 0;
+  }
+
+  api = nr_php_get_class_constant(ce, "API");
+  if (NULL == api) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s: Logger does not have API", __func__);
+    return 0;
+  }
+
+  if (nr_php_is_zval_valid_integer(api)) {
+    retval = Z_LVAL_P(api);
+  } else {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "%s: expected API be a valid int, got type %d", __func__,
+                     Z_TYPE_P(api));
+  }
+
+  nr_php_zval_free(&api);
+  return retval;
+}
+
+/*
+ * Purpose : Convert $level argument of Monolog\Logger::addRecord to a string
+ * representation of Monolog's log level.
+ *
+ * Params  : Logger instance and Monolog\Logger::addRecord argument list
+ *
+ * Returns : A new string with Monolog's log level name
+ */
+static char* nr_monolog_get_level_name(zval* logger,
+                                       NR_EXECUTE_PROTO TSRMLS_DC) {
+  zval* level_name = NULL;
+  char* level_name_string = nr_strdup("UNKNOWN");
+
+  if (!nr_php_object_has_method(logger, "getLevelName" TSRMLS_CC)) {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "%s: Logger does not have getLevelName method", __func__);
+    return level_name_string;
+  }
+
+  zval* level = nr_php_arg_get(1, NR_EXECUTE_ORIG_ARGS TSRMLS_CC);
+  level_name = nr_php_call(logger, "getLevelName", level);
+  if (nr_php_is_zval_valid_string(level_name)) {
+    level_name_string = nr_strdup(Z_STRVAL_P(level_name));
+  } else {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "%s: expected level_name be a valid string, got type %d",
+                     __func__, Z_TYPE_P(level_name));
+  }
+
+  nr_php_zval_free(&level_name);
+  nr_php_arg_release(&level);
+
+  return level_name_string;
+}
+
+/*
+ * Purpose : Convert $message argument of Monolog\Logger::addRecord to a string
+ *
+ * Params  : Monolog\Logger::addRecord argument list
+ *
+ * Returns : A new string with Monolog's log message; caller must free
+ */
+
+static char* nr_monolog_get_message(NR_EXECUTE_PROTO TSRMLS_DC) {
+  char* message = NULL;
+  zval* message_arg = NULL;
+
+  message_arg = nr_php_arg_get(2, NR_EXECUTE_ORIG_ARGS TSRMLS_CC);
+  if (!nr_php_is_zval_valid_string(message_arg)) {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "%s: expected $message be a valid string, got type %d",
+                     __func__, Z_TYPE_P(message_arg));
+    message = nr_strdup("");
+  } else {
+    message = nr_strdup(Z_STRVAL_P(message_arg));
+  }
+  nr_php_arg_release(&message_arg);
+
+  return message;
+}
+
+/*
+ * Purpose : Format key of $context array's element as string
+ *
+ * Params  : zend_hash_key
+ * *
+ * Returns : A new string representing zval; caller must free
+ *
+ */
+static char* nr_monolog_fmt_context_key(zend_hash_key* hash_key) {
+  char* key_str = NULL;
+  zval* key = nr_php_zval_alloc();
+  if (nr_php_zend_hash_key_is_string(hash_key)) {
```

Comment:
Is it safer to use `nr_php_is_zval_non_empty_string(hash_key)` ?

Reply:
`hash_key` is not of `zval` type so it's not possible to use `nr_php_is_zval_non_empty_string`. However `nr_php_zend_hash_key_is_string` does check if the `hash_key` is NULL and returns false in such case

https://github.com/newrelic/newrelic-php-agent/blob/9fd43b56a296e761156db9aa08a1919a975cd56c/agent/php_hash.h#L28

This behavior ensures that this function returns `unsuported-key-type` generated in line 147.